logstash-core/src/main/java/org/logstash/ackedqueue/Page.java

```diff
@@ -99,10 +98,9 @@ public boolean isFullyRead() {
     }
 
     public boolean isFullyAcked() {
-        // TODO: it should be something similar to this when we use a proper bitset class like ES
-        // this.ackedSeqNum.firstUnackedBit >= this.elementCount;
-        // TODO: for now use a naive & inefficient mechanism with a simple Bitset
-        return this.elementCount > 0 && this.ackedSeqNums.cardinality() >= this.elementCount;
+        final int cardinality = ackedSeqNums.cardinality();
+        return elementCount > 0 && cardinality == ackedSeqNums.length()
```

Comment:
Adjusted this method, it was wrong before! Just checking the cardinality isn't enough here, we need to make sure that cardinality and length are the same. If they aren't then we're in a broken state.

Reply:
Seems like we need an assertion here to check for invariants. Do we need to assert that `ackedSeqNums.cardinality() <= this.elementCount`?

Reply:
@andrewvc that's basically impossible because of the assertions we have in `#ack(` itself => I'll leave it out for now. I'll add a follow up PR to this one today that has additional simplifications here, then we can revisit if you still feel like we need this one :) I'll assign you there for review :)